src/dune_pkg/repository_id.ml

```diff
@@ -0,0 +1,83 @@
+open Stdune
+module Process = Dune_engine.Process
+module Display = Dune_engine.Display
+
+type t =
+  | Git_hash of string
+  | Opam_stamp of string
+
+let equal a b =
+  match a, b with
+  | Git_hash a, Git_hash b | Opam_stamp a, Opam_stamp b -> String.equal a b
+  | _, _ -> false
+;;
+
+let encode : t -> Dune_sexp.t = function
+  | Git_hash commitish ->
+    List
+      [ Dune_lang.atom_or_quoted_string "Git_hash"
+      ; Dune_lang.atom_or_quoted_string commitish
+      ]
+  | Opam_stamp stamp ->
+    List
+      [ Dune_lang.atom_or_quoted_string "Opam_stamp"
+      ; Dune_lang.atom_or_quoted_string stamp
+      ]
+;;
+
+let decode =
+  let open Dune_sexp.Decoder in
+  let+ constr, stamp = pair string string in
+  match constr with
+  | "Git_hash" -> Git_hash stamp
+  | "Opam_stamp" -> Opam_stamp stamp
+  | _ -> failwith "TODO: parsing failure"
+;;
+
+let to_dyn = function
+  | Git_hash commitish -> Dyn.Variant ("Git_hash", [ String commitish ])
+  | Opam_stamp stamp -> Dyn.Variant ("Opam_stamp", [ String stamp ])
+;;
+
+let repo_id_of_stamp dir =
+  let local = Path.Local.of_string "repo" in
+  let repo_file = Path.append_local dir local in
+  match Path.exists repo_file with
+  | false ->
+    (* no repo file, we can't read this, bail *)
+    Fiber.return None
+  | true ->
+    (match OpamParser.FullPos.file (Path.to_string repo_file) with
+     | exception Stdlib.Parsing.Parse_error -> Fiber.return None
+     | opam_file ->
+       let stamp =
+         List.find_map opam_file.file_contents ~f:(fun { pelem; _ } ->
+           match pelem with
+           | Variable ({ pelem = "stamp"; _ }, { pelem = String stamp; _ }) -> Some stamp
+           | _ -> None)
+       in
+       stamp |> Option.map ~f:(fun s -> Opam_stamp s) |> Fiber.return)
+;;
+
+let of_path dir =
+  let open Fiber.O in
+  let failure_mode = Process.Failure_mode.Return in
+  let display = Display.Silent in
```

Comment:
What's the difference between Quiet and Silent?

Reply:
`Quiet` still displays error messages when commands fail (and they output an error message), in our case that's a lot of `fatal: invalid gitfile format: $DIR/dune/_build/.sandbox/.git`. I considered changing the definition of `Quiet` but wasn't sure if that wouldn't break a lot of places that depended on printing the error message.

Reply:
Note that this error is intentional and that your test is trying to escape the sandbox (by running git without `-C` or without initializing a repo in the test)

Reply:
The error in this case is intentional since the repo is not a valid repo, so it should fall back to use the `repo` file (which it does). But with `Quiet` it will always display this error message before falling back to the other code, which I would say is overly alarming.

I can either set `-C` or `~dir` to the folder, either way the result is the same. I don't think setting `-C` makes it not escape the sandbox. The git documentation makes it sound like it's basically `cd ~dir; git ...`.

However I think that maybe implementing `git rev-parse HEAD` is easy/stable enough to not use the `git` binary for it.

Reply:
You could also just do `git init` inside the test.

Reply:
I am, in line 136?

Reply:
Ah, now I understand, git traverses up and then hits that (deliberate trap) folder and breaks. It looks like this can be disabled using `--git-dir`.